### ISSUE TITLE
0 byte mp4 file after recording

### ISSUE THREAD

**Issue by aryasenna** [phw/peek#130]
Hi, 
I was wondering why mp4 recording isn't working on Fedora 25 + peek-1.0.1-2.1.x86_64 + X.Org
I have installed ffmpeg.

GIF and WebM works fine, but when I choose mp4 it only results in 0 byte mp4 file.

How exactly I can debug this? tried to run in via terminal, but no apparent error:

> [me@localhost ~] $ peek
Using screen recorder backend gnome-shell
Recording to file /home/me/.cache/peek/peekHSSXYY.mp4

It appears the rendering process doesn't take place at all.

**Comment by phw** [comment]
By default if you are running on Gnome Peek will use the gnome-shell backend. That requires the gstreamer bad plugins for MP4 recording. The gstreamer1-plugins-bad-freeworld package from the RPM fusion repository provides those.

Alternatively you can use the ffmpeg backend by launching Peek with 

    peek -b ffmpeg

I should add a note about this to the README. And the main issue here is better error reporting in Peek, but that is planned for one of the next releases :)

**Comment by aryasenna** [comment]
Okay, but I have `gstreamer1-plugins-bad-freeworld` installed already.


Anyhow, adding the '-b ffmpeg` works but with awarning:
`[swscaler @ 0x1292900] Warning: data is not aligned! This can lead to a speedloss`

I guess I'll modify the desktop file and set it re-modify it on every update for now.
Thanks! I think it's fine to close the issue?

**Comment by phw** [comment]
Strange that you have that package installed but it doesn't work, maybe the MP4 support requires yet another package. Could you do me a favour and look at the output of "journalctl -f" when you start recording with the gnome-shell? This could give us some clue why it fails.

**Comment by aryasenna** [comment]
Sure, here you go.
```
[me@localhost ~] $ peek
Using screen recorder backend gnome-shell
Recording to file /home/aryasenna/.cache/peek/peekWBKOYY.mp4
^C
[me@localhost ] $ journalctl -f
Apr 16 15:23:31 localhost org.gnome.Shell.desktop[5568]: Recording to /home/me/.cache/peek/peekWBKOYY.mp4
Apr 16 15:23:31 localhost gnome-shell[5568]: Error in recording pipeline: Can not initialize x264 encoder.
Apr 16 15:23:31 localhost gnome-shell[5568]: recorder_record_frame: assertion 'recorder->current_pipeline != NULL' failed
Apr 16 15:23:33 localhost gnome-shell[5568]: shell_recorder_close: assertion 'recorder->state != RECORDER_STATE_CLOSED' failed
```

**Comment by phw** [comment]
Thanks. I cannot tell right away if the message "Can not initialize x264 encoder" means the x264 encoder is not installed or something is broken with it. I will do some testing with Fedora myself and try to figure out what needs to be installed for this to run and update the README

**Comment by aryasenna** [comment]
You're welcome. I'll let the issue open, till then, I guess.

**Comment by phw** [comment]
Looked at this a bit, there are two issues here:

1. The documentation and package dependencies are wrong, you need gstreamer1-plugins-ugly installed as this contains the x264 encoder. I have checked in a fix for this.
2. In your case you seem to already have this installed, but I can reproduce the "Error in recording pipeline: Can not initialize x264 encoder." error on my Fedora installation.

This second issue is a bit critical, as I experimented with also using the x264 encoder for GIF creation under gnome-shell to fix for #116. Trying to figure out what is wrong on Fedora and whether I can workaround it or document a solution.

**Comment by phw** [comment]
x264 encoding seems to be broken on Fedora. I tried it from command line with:

    gst-launch-1.0 -v filesrc location=test.webm ! decodebin ! x264enc ! mp4mux ! filesink location=output.mp4

Same thing works when using WebM:

    gst-launch-1.0 -v filesrc location=test.webm ! decodebin ! vp8enc ! webmmux ! filesink location=output.webm

**Comment by aryasenna** [comment]
Interesting. I am not familiar with the x264 encoder and what package it contains to file a bug report. But I recall "ugly" means that package is not distributed by Fedora Projects. A quick Google search seems to point into rpmfusion (http://download1.rpmfusion.org/free/fedora/releases/25/Everything/x86_64/os/repoview/gstreamer-plugins-ugly.html), and FWIW I used the one from negativo17: https://github.com/negativo17/gstreamer1-plugins-ugly

Perhaps before the x264 encoder is fixed, as a workaround Fedora build should depend on ffmpeg by default?

**Comment by phw** [comment]
Yes, RPM Fusion provides it.

> Perhaps before the x264 encoder is fixed, as a workaround Fedora build should depend on ffmpeg by default?

As Fedora defaults to GNOME on Wayland this is not really a solution. But maybe I could default to the gnome-shell backend only if it detects both a GNOME session and Wayland. Currently it checks only for GNOME and enables it there.

**Comment by phw** [comment]
I tested my above command lines with test video files from http://techslides.com/sample-webm-ogg-and-mp4-video-files-for-html5 and then even the x264 encoder worked. But if used with the gnome-shell DBus service or when trying to decode and re-encode a webm file generated by Peek it fails. This seems to indicate that while there seems to be some bug with the encoder in Fedora 25 it might be possible to get a workaround for it.

**Comment by phw** [comment]
I wonder whether this is actually  duplicate of https://github.com/phw/peek/issues/141 . I will test again with newest release

**Comment by aryasenna** [comment]
Sorry I have been awhile for quite a while. Just upgraded to Fedora 26 and this issue is no longer reproducible. Not sure if's due to change in Fedora or fix on #141 as you suggested.

So I think it's safe to close this issue..?

P.s. any plan to make Peek works in Wayland?

**Comment by phw** [comment]
> P.s. any plan to make Peek works in Wayland?

Sorry for the late answer on this, missed this comment. Peek already works on Wayland in a Gnome session (will use the Gnome Shell DBus API for screen recording). Requires XWayland to be present (but that should be the default). See https://github.com/phw/peek#why-no-native-wayland-support for details.

For this issue I close it, might very well have been #141
